File: benchmarks/project/src/geam_benchmarks_ffi.mjs
import { Result$Error, Result$Ok } from "./gleam.mjs";
import { hrtime } from "node:process";

const origin = hrtime.bigint();

export function environment(name) {
  const value = process.env[name];
  return value === undefined
    ? Result$Error(`Missing environment variable: ${name}`)
    : Result$Ok(value);
}

export function monotonic_ns() {
  return Number(hrtime.bigint() - origin);
}

export function consume(value) {
  globalThis.geamBenchmarkResult = value;
  return value;
}

File: benchmarks/project/src/geam_benchmarks_ffi.erl
-module(geam_benchmarks_ffi).
-export([environment/1, monotonic_ns/0, consume/1]).

environment(Name) ->
    case os:getenv(binary_to_list(Name)) of
        false -> {error, <<"Missing environment variable: ", Name/binary>>};
        Value -> {ok, unicode:characters_to_binary(Value)}
    end.

monotonic_ns() -> erlang:monotonic_time(nanosecond).

consume(Value) ->
    erlang:put(geam_benchmarks_consumed, Value),
    Value.
